governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -133,33 +143,60 @@ public void run() {
 
     }
 
-    private Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    private Map<Object, UnscopedCleanupAction> cleanupActions = new WeakHashMap<>();
     private ScopeCleaner scopeCleaner = new ScopeCleaner();
-
-    Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private Map<Class<? extends Annotation>, Scope> scopeBindings;
 
     public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
         this.scopeBindings = new HashMap<>(scopeBindings);
     }
 
     public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
```

Comment:
Looks like we check externally that `action` is not empty but might be good to either add a precondition check or return false if `action` is empty.

Reply:
can't do it with an Iterable without getting the iterator() first :(